rust/src/io/commit.rs

```diff
@@ -263,19 +269,227 @@ impl<T: CommitLock + Send + Sync + Debug> CommitHandler for T {
+    )
+    .await?;
+
+    Ok(manifest)
+}
+
+/// Attempt to commit a transaction, with retries and conflict resolution.
+pub(crate) async fn commit_transaction(
+    dataset: &Dataset,
+    object_store: &ObjectStore,
+    transaction: &Transaction,
+    write_config: &ManifestWriteConfig,
+    commit_config: &CommitConfig,
+) -> Result<Manifest> {
+    // Note: object_store has been configured with WriteParams, but dataset.object_store()
+    // has not necessarily. So for anything involving writing, use `object_store`.
+    let transaction_file = write_transaction_file(object_store, &dataset.base, transaction).await?;
+
+    let mut dataset = dataset.clone();
+    // First, get all transactions since read_version
+    let mut other_transactions = Vec::new();
+    let mut version = transaction.read_version;
+    loop {
+        version += 1;
+        match dataset.checkout_version(version).await {
+            Ok(next_dataset) => {
+                let other_txn = if let Some(txn_file) = &next_dataset.manifest.transaction_file {
+                    Some(read_transaction_file(object_store, &next_dataset.base, txn_file).await?)
+                } else {
+                    None
+                };
+                other_transactions.push(other_txn);
+                dataset = next_dataset;
+            }
+            Err(crate::Error::NotFound { .. }) | Err(crate::Error::DatasetNotFound { .. }) => {
+                break;
+            }
+            Err(e) => {
+                return Err(e);
+            }
+        }
+    }
+
+    let mut target_version = version;
+
+    // If any of them conflict with the transaction, return an error
+    for (version_offset, other_transaction) in other_transactions.iter().enumerate() {
+        let other_version = transaction.read_version + version_offset as u64 + 1;
+        check_transaction(transaction, other_version, other_transaction)?;
+    }
+
+    for _ in 0..commit_config.num_retries {
+        // Build an up-to-date manifest from the transaction and current manifest
+        let (mut manifest, indices) = transaction.build_manifest(
+            Some(dataset.manifest.as_ref()),
+            dataset.load_indices().await?,
+            &transaction_file,
+            write_config,
+        )?;
+
+        debug_assert_eq!(manifest.version, target_version);
+
+        // Try to commit the manifest
+        let result = write_manifest_file(
+            object_store,
+            &dataset.base,
+            &mut manifest,
+            if indices.is_empty() {
+                None
+            } else {
+                Some(indices.clone())
+            },
+            write_config,
+        )
+        .await;
+
+        match result {
+            Ok(()) => {
+                return Ok(manifest);
+            }
+            Err(CommitError::CommitConflict) => {
+                // See if we can retry the commit
+                dataset = dataset.checkout_version(target_version).await?;
```

Comment:
minor: multiple version could have passed here, we should pull till the latest one.

Let's just add a ticket for this? No need to fix it right now.

Reply:
https://github.com/lancedb/lance/issues/1139